packages/core/src/types.ts

```diff
@@ -1854,21 +1856,36 @@ export type EventOfMachine<
 export interface ActorContext<TEvent extends EventObject, TSnapshot> {
   self: ActorRef<TEvent, TSnapshot>;
   name: string;
-  _event: SCXML.Event<TEvent> | LifecycleSignal;
+  sessionId: string;
+  logger: (...args: any[]) => void;
+  exec: ((fn: () => void) => void) | undefined;
+  defer: ((fn: () => void) => void) | undefined;
+  observers: Set<Observer<TSnapshot>>;
 }
 
-export interface Behavior<TEvent extends EventObject, TSnapshot = any> {
+export interface Behavior<
+  TEvent extends EventObject,
+  TSnapshot = any,
+  TInternalState = any
+> {
   transition: (
-    state: TSnapshot,
+    state: TInternalState,
     message: TEvent | LifecycleSignal,
     ctx: ActorContext<TEvent, TSnapshot>
-  ) => TSnapshot;
-  initialState: TSnapshot;
-  start?: (actorCtx: ActorContext<TEvent, TSnapshot>) => TSnapshot;
+  ) => TInternalState;
+  // TODO: should we only use `getInitialState(...)`?
```

Comment:
Yeah, we should :P

Reply:
This is a little tricky to do; will save for follow-up PR